lib/rack/lint.rb

```diff
@@ -356,6 +356,15 @@ def check_environment(env)
         unless env[SCRIPT_NAME] != "/"
           raise LintError, "SCRIPT_NAME cannot be '/', make it '' and PATH_INFO '/'"
         end
+
+        ## <tt>rack.response_finished</tt>:: An array of callables run after the HTTP response has been sent.
+        if callables = env[RACK_RESPONSE_FINISHED]
+          raise LintError, "rack.response_finished must be an array of callable objects" unless callables.is_a?(Array)
+
+          callables.each do |callable|
+            raise LintError, "rack.response_finished values must respond to call" unless callable.respond_to?(:call)
```

Comment:
I was originally checking `.arity` here too to ensure it's <= 1 but that seemed like it may needlessly rigid since now objects that respond to `call` can be passed as well.

It does bring up a question about errors though. If an exception is raised do we want servers to pass it to `call`, or do we leave it up to the server to check for arity and pass the exception if it accepts an argument?

Reply:
Regarding arity, I think one way to handle this is to wrap the array with a linter which validates the calls and responses.

Regarding errors, I think we need to make a decision about this:

1. Either we care about errors which occur when writing the response, and provide a mechanism for dealing with them via this interface.
2. We ignore errors or provide some other interface for receiving errors.

The difference between (1) and (2) is that (2) is less well specified and thus server implementations have more flexibility as to how they implement the specification.

Reply:
> Regarding arity, I think one way to handle this is to wrap the array with a linter which validates the calls and responses.

I think I see what you mean, but let me know if this isn't correct. The idea is to make sure each value is called and doesn't raise? There's no specific return value, at least in the current implementation of `rack.after_reply` which I'm basing this on.

> Regarding errors, I think we need to make a decision about this:
> 
> 1. Either we care about errors which occur when writing the response, and provide a mechanism for dealing with them via this interface.
> 2. We ignore errors or provide some other interface for receiving errors.
> 
> The difference between (1) and (2) is that (2) is less well specified and thus server implementations have more flexibility as to how they implement the specification.

Yeah, I'm on the fence here. The implementations I've seen/implemented myself don't need to check for errors since the code would have been run anyways, but I can imagine there's at least a few use-cases where it would be helpful to have the error.

If we omit it from the spec, I think there's a few different ways the error could be retrieved. One thought would be to catch the exception in a middleware and store it in the env since the callable may have access to the env.

Reply:
> I think I see what you mean, but let me know if this isn't correct. The idea is to make sure each value is called and doesn't raise? There's no specific return value, at least in the current implementation of rack.after_reply which I'm basing this on.

Yes, wrap the values with an internal validation layer.

I don't think I mind what you do here, but lets define something clearly in the spec w.r.t. how well behaved code should execute.

Let's also make it clear when we generally expect execution to occur (in the spec).